fdbserver/storageserver.actor.cpp

```diff
@@ -1114,22 +1121,31 @@ ACTOR Future<GetKeyValuesReply> readRange( StorageServer* data, Version version,
 
 		while (limit>0 && *pLimitBytes>0 && readBegin < range.end) {
 			// ASSERT( vStart == view.lower_bound(readBegin) );
-			ASSERT( !vStart || vStart.key() >= readBegin );
-			if (vStart) { auto b = vStart; --b; ASSERT( !b || b.key() < readBegin ); }
+			ASSERT( !vCurrent || vCurrent.key() >= readBegin );
 			ASSERT( data->storageVersion() <= version );
 
-			// Read up to limit items from the view, stopping at the next clear (or the end of the range)
-			vEnd = vStart;
-			vCount = 0;
-			int vSize = 0;
-			while (vEnd && vEnd.key() < range.end && !vEnd->isClearTo() && vCount < limit && vSize < *pLimitBytes){
-				vSize += sizeof(KeyValueRef) + vEnd->getValue().expectedSize() + vEnd.key().expectedSize();
-				++vCount;
-				++vEnd;
+			/* Traverse the PTree further, if thare are no unconsumed resultCache items */
+			if (pos == resultCache.size()) {
+				if (vCurrent) {
+					auto b = vCurrent;
+					--b;
+					ASSERT(!b || b.key() < readBegin);
+				}
+
+				// Read up to limit items from the view, stopping at the next clear (or the end of the range)
+				int vSize = 0;
+				while (vCurrent && vCurrent.key() < range.end && !vCurrent->isClearTo() && vCount < limit &&
+					   vSize < *pLimitBytes) {
+					// Store the versionedData results in resultCache
+					resultCache.push_back(result.arena, KeyValueRef(vCurrent.key(), vCurrent->getValue()));
```

Comment:
It is a little weird to be be using result.arena to hold the resultCache vector, although maybe okay because the result should be short lived because we will be sending it over the network and freeing the arena pretty shortly after creating it.

Reply:
Yes, that was the intent. I started out with a different Arena, but changed to result.arena on Markus's suggestion because of it's lifetime.